align_system/algorithms/oracle_adm.py

```diff
@@ -0,0 +1,321 @@
+
+log = logging.getLogger(__name__)
+
+
+class OracleADM(ActionBasedADM):
+    def __init__(self, probabilistic: bool=False, upweight_missing_kdmas: bool=False, **kwargs):
+        self.probabilistic = probabilistic
+        self.upweight_missing_kdmas = upweight_missing_kdmas
+
+    def _inverse_distance(self, target_kdma_values, system_kdmas, min_value=0., max_value=1.):
+        """
+        Compute 1/(Euclidean distance between the target and system KDMAs)
+        """
+        if system_kdmas is None:
+            system_kdmas = dict()
+
+        distance = 0.
+
+        # TODO: Determine if this is an appropriate measure
+        for kdma, target in target_kdma_values.items():
+            # Edge case #1: Target KDMA is not present in System KDMAs, assign "midpoint" distance
+            # The idea here is that an unlabeled choice may be more desired than a fully misaligned one
+            if kdma not in system_kdmas:
+                val_range = max_value - min_value
+                midpoint = val_range/2. + min_value
+                if self.upweight_missing_kdmas:
+                    if self.probabilistic:
+                        midpoint += random.uniform(-val_range, val_range)
+                    distance += midpoint**2
+                else:
+                    distance +=  (midpoint + abs(target - midpoint) + 0.1)**2  # Assign maximum distance
+            else:
+                distance += (target - system_kdmas[kdma])**2
+
+        #Edge case #2: System KDMA is not present in Target KDMAs => ignoring for now
+        # e.g if Target: X=10; Choice A: X=3; Choice B: X=3, Y=5; then A and B are equivalent
+
+        # TODO: Should we just return 1/distance**2?
+        # Small epsilon for a perfect (0 distance) match
+        return math.sqrt(1/(distance+1e-16))
+
+    def choose_action(self, scenario_state, available_actions, alignment_target, **kwargs):
+        if not available_actions:
+            return None
+
+        if alignment_target is None:
+            raise ValueError("Oracle ADM needs alignment target")
+
+        # Build out the corresponding kdma_association target dictionary
+        target_kdma_assoc = {
+            target['kdma']: target['value']
+            for target in alignment_target.kdma_values
+        }
+
+        # Weight action choices with distance-based metric
+        inv_dists = [
+            self._inverse_distance(target_kdma_values=target_kdma_assoc, system_kdmas=action.kdma_association)
+            for action in available_actions
+        ]
+
+        # Convert inverse distances to probabilities
+        probs = [inv_dist/sum(inv_dists) for inv_dist in inv_dists]
+
+        if self.probabilistic:
+            action_to_take = np.random.choice(available_actions, p=probs)
+        else:  # Always choose max probability action (chooses first if multiple are tied)
+            action_to_take = available_actions[probs.index(max(probs))]
```

Comment:
Tagging @brianhhu here to be sure, but I think Brian had intended the probabilistic piece here to be a knob to turn (something like 95% of the time it picks the best answer vs. 90% vs. 80% etc.) rather than use the distance from the target as probabilistic weights.  Brian could you confirm?

Reply:
Yeah, the probabilistic piece needs the most work still. Happy to hear any feedback/desires there

I thought Brian mentioned only smoothing probabilities without a matching KDMA, but in the current scenarios that doesn't really affect the final outcome (as actions without KDMAs are essentially no-ops right now). My "smoothing" technique is also not very technical right now, so that could likely be improved, just not sure how

Reply:
Again will wait for Brian to chime in here, but perhaps a simple way to implement what I was suggesting above would be to check `random.random() < threshold` (where `threshold` is `0.95` or `0.9` etc.) and if the check passes just pick the "right" answer, otherwise just pick a random "not right" answer.

Reply:
The idea of the probabilistic action sampling is that it can potentially support a wide variety of decision strategies (all within the same framework):

1) random sampling would just be uniform distribution of probabilities over all actions
2) the "greedy" approach would just put all the probability mass on one action (i.e. "most aligned" given the target KDMA), or a small subset actions if there's a tie
3) in between the random/greedy strategies would be sampling probabilities of actions weighted by some distance from the target KDMA- this allows for some natural variability in decisions (approximating humans), where it's possible for a "suboptimal" decision to be chosen on a given probe (but on average, decisions would be aligned)

I'll go through the code in a bit more detail to see if I understand what's going on and can provide additional feedback then.

Reply:
I removed this `self.probabalistic` if condition in the latest, the portion of the code that's most relevant to this discussion is probably [here](https://github.com/ITM-Kitware/align-system/pull/46/files#diff-717c95a115392c6e4474060165d525403aa1a6ec36c071483540ab1b1d179ccdR75-R80)

1. I think this would just be the `RandomADM`, correct me if I'm wrong
2. I believe this is the `else` portion of the linked code
3. I believe this is the `if` portion of the linked code. I'm not sure if we need to add additional weight to non-optimal choices to get more probability of taking those paths.[ Upweighting missing KDMAs](https://github.com/ITM-Kitware/align-system/pull/46/files#diff-717c95a115392c6e4474060165d525403aa1a6ec36c071483540ab1b1d179ccdR39-R40) may or may not be enough for that

Reply:
Sorry I lost track of this MR, and thanks for addressing my last comment. I don't see a drawback for merging, as @eveenhuis reported alignment scores in-line with the maximum possible reported by TA1. We should go back and also check the Parallax "maximally aligned" metric (which should be close to 100%), but that will require generating the output files using this ADM.